### ISSUE TITLE
Discrete spaces, sampling stuck

### ISSUE THREAD

**Issue by segasai** [scikit-optimize/scikit-optimize#781]
Hi, 

I'm using skopt to optimize  a function with discrete arguments and  the optimizer seems to be gett getting stuck.
Here is weird illustration of the problem. 
When I ask input points from the optimizer, instead of expecting uniform distributions, I'm  getting corners of the distribution, see example below

```
from skopt import Optimizer
from skopt.space import Real,Integer

optimizer = Optimizer(
            dimensions=[Integer(-40,40), Integer(-40,40)],
                    random_state=1  )
print (optimizer.ask(n_points=100))
>>> [[16, -35], [-10, -28], [-1, 31], [-35, -20], [-2, 14], [-15, 37], [-38, 35], [26, -29], [13, 10], [33, 0], [-39, -40], [40, 40], [40, -40], [-39, 40], [-40, -40], [40, 40], [40, -39], [-40, 40], [40, 39], [-39, -40], [40, -40], [-40, 39], [-40, -40], [40, 40], [40, -40], [-40, 40], [-40, -40], [40, 40], [40, -40], [-40, 40], [-40, -39], [40, 40], [40, -40], [-40, 40], [-40, -40], [40, 39], [40, -40], [-40, 39], [-40, -40], [40, 39], [40, -39], [-40, -39], [40, 40], [-40, 39], [40, -40], [-40, 40], [-40, -39], [40, 40], [40, -39], [-40, 40], [40, 40], [-39, -40], [40, -40], [-40, 40], [-40, -40], [40, 39], [-40, 40], [40, -39], [40, 40], [40, -40], [-40, -40], [-40, 40], [39, 40], [-40, -40], [40, -40], [-39, 40], [-40, -40], [40, 40], [-40, 39], [39, -40], [40, -40], [-40, -39], [40, 40], [-40, 40], [40, 39], [-38, -40], [40, -39], [-40, 40], [40, -39], [40, 40], [-40, -40], [-40, 39], [39, 40], [-40, -40], [40, 40], [39, -40], [-40, -40], [-40, 40], [40, -40], [-40, 40], [-40, -40], [40, -40], [40, 40], [-40, -40], [-40, 40], [40, -40], [40, 40], [-40, 39], [-40, -40], [40, 39]]

```
Surely that doesn't look right.

Thanks 
         Sergey

**Comment by jredondopizarro** [comment]
Maybe the corners of the distribution are the values that minimize your objective function?

**Comment by segasai** [comment]
Here there is no even an objective function, so I'd expect proposed points to be just uniformly distributed over parameter space.

**Comment by jredondopizarro** [comment]
Yes, you are right, I had not realized that. I have tested your code myself and I obtain the same results, this surely is some kind of malfunction.

**Comment by sqbl** [comment]
Hi,

I do not believe that these are errors. 
When you are requesting 100 sets of parameters during your .ask() it will return one new point at a time - then it internally lies to itself regarding the value of the n'th point before asking for the n+1'th point. The documentation of the code links to this paper: ([https://hal.archives-ouvertes.fr/hal-00732512/document].
Now, without any data at all I assume that the constant liar will return a point from the prior - which might (?) be an uniform distribution - hence it will assume that the entire function-surface is flat. In the end, when everything is flat the outerpoints will appear to have the largest uncertainty and the acquisition function will suggest experiments here.
If you really want to suggest 100 points at a time, I guess that you plan on running the calculations in parallel? Bayesian Optimization is really strong in optimizing expensive black box cases that are best run in sequence - to ensure full learning from each experiment before the next is conducted. So if you can run 100 experiments at a time you might want to consider an old, tried and tested approach such as DoE.

Hope it helps

**Comment by segasai** [comment]
Thank you for the opinion. I do disagree however. No matter how the code is implemented, I just don't see how any output that at certain point is just stuck at the corner can make sense. This function should return uniform distribution from the prior (IMO). 

And yes. I was trying to parallelize the optimization. I have a slow single threaded function that I was trying to optimise.

In the end I didn't manage to make skopt do anything meaningful for my purely discrete problem, and I had to switch to gpyopt... Maybe I was doing something wrong, but in my opinion, something is not quite right for discrete problems, which is illustrated by the issue that I raised.

**Comment by sqbl** [comment]
Good to hear that you got a solution to your problem. 
For future readers, I would like to refer to this paper, that deals with boundary over-exploration: https://arxiv.org/pdf/1704.00963
Furthermore, If you are interested in random sampling of the prior, then perhaps a latin hypercube sampling (or a SOBOL) as discussed in issue #433 could be relevant?

Best
Søren

**Comment by chnzhangrui** [comment]
Hi @segasai,

I believe what you saw is expected. When you define `Optimizer`, you should also provide `n_initial_points=100` if you want to do a random search on the first 100 points.
In your case, the default value of `n_initial_points=10` is used, therefore, after the 10 points, it will start giving you points based on the surrogate function. Since you didn't tell it anything, most likely it will give you the points at corners, as you observed.

In short, use the following:
`optimizer = Optimizer(
            dimensions=[Integer(-40,40), Integer(-40,40)],
                    random_state=1, n_initial_points=100 )`
It should work.

**Comment by nurettin** [comment]
@chnzhangrui 

If you simplify,

```
    optimizer = Optimizer(dimensions=[Integer(1, 10)],
                          random_state=1, n_initial_points=10)
    print(optimizer.ask(n_points=10))
```

The result repeats.

`[[7], [2], [4], [2], [5], [9], [2], [3], [5], [7]]`

@sqbl 
If you use lhs

```
    optimizer = Optimizer(dimensions=[Integer(1, 10)],
                          random_state=1, n_initial_points=10, initial_point_generator="lhs")
    print(optimizer.ask(n_points=10))
```

It doesn't repeat

`[[7], [10], [4], [2], [5], [8], [6], [9], [1], [3]]`

But if we increase the number of points

```
    optimizer = Optimizer(dimensions=[Integer(1, 20)],
                          random_state=1, n_initial_points=20, initial_point_generator="lhs")
    print(optimizer.ask(n_points=20))
```

`[[12], [6], [8], [7], [9], [18], [16], [10], [19], [19], [14], [6], [2], [13], [4], [5], [14], [11], [2], [16]]`

Now it repeats (you can see 2, 19 and 16 doubles, 1 is missing etc.)

You might as what the point of all this is. It is a good question. Sometimes people just want to scan the entire space and we should be able to do that if "grid" sampling is selected while using the "DUMMY" optimizer so that we don't have to rewrite the code to do a brute force for all parameters.
